File: autokeren/tools/rewind.py
"""Rewind tool — undo N tool call terakhir via checkpoint system."""
from __future__ import annotations

from typing import Any

from autokeren.checkpoints.manager import CheckpointManager
from autokeren.tools.base import Tool, ToolResult


class RewindTool(Tool):
    name = "rewind"
    description = (
        "Undo N tool call terakhir. Restore file state ke checkpoint sebelumnya. "
        "Gunakan setelah agent ngerusak codebase atau mau coba pendekatan berbeda."
    )
    parameters = {
        "type": "object",
        "properties": {
            "steps": {
                "type": "integer",
                "description": "Jumlah tool call yang di-undo (default: 1)",
                "default": 1,
            },
        },
    }

    def __init__(self, manager: CheckpointManager) -> None:
        self.manager = manager

    def run(self, steps: int = 1, **_: Any) -> ToolResult:
        total = self.manager.count()
        if total == 0:
            return ToolResult(output="Tidak ada checkpoint untuk di-rewind.")
        actual = min(steps, total)
        undone = self.manager.rewind(actual)
        if not undone:
            return ToolResult(output="Tidak ada checkpoint untuk di-rewind.")
        lines = [f"⏪ Rewind {len(undone)} tool call:\n"]
        for cp in undone:
            tool = cp.tool_name
            arg_path = cp.tool_args.get("path", cp.tool_args.get("query", ""))
            n_changes = len(cp.file_changes)
            lines.append(f"  #{cp.id} {tool}({arg_path}) — {n_changes} file di-revert")
        remaining = self.manager.count()
        lines.append(f"\nCheckpoint tersisa: {remaining}.")
        return ToolResult(output="\n".join(lines))

File: internal/tool/rewind.go
package tool

import (
	"context"
	"fmt"
	"strings"

	"github.com/autokeren/autokeren/internal/checkpoint"
)

type Rewind struct{ Manager *checkpoint.Manager }

func (r Rewind) Definition() Definition {
	return Definition{Name: "rewind", Description: "Kembalikan perubahan file dari checkpoint terakhir.", RequiresPermission: true, Parameters: map[string]any{"type": "object", "properties": map[string]any{"steps": map[string]any{"type": "integer"}}}}
}

func (r Rewind) NeedsPermission(map[string]any) (bool, string) {
	return true, "kembalikan perubahan file dari checkpoint"
}

func (r Rewind) Run(ctx context.Context, args map[string]any, _ Emitter) Result {
	select {
	case <-ctx.Done():
		return Result{OK: false, Error: ctx.Err().Error()}
	default:
	}
	if r.Manager == nil {
		return Result{OK: false, Error: "time-travel tidak diaktifkan"}
	}
	steps := numericArg(args["steps"], 1)
	undone, err := r.Manager.Rewind(steps)
	if err != nil {
		return Result{OK: false, Error: err.Error()}
	}
	if len(undone) == 0 {
		return Result{OK: true, Output: "Tidak ada checkpoint untuk di-rewind."}
	}
	lines := []string{fmt.Sprintf("⏪ Rewind %d tool call:", len(undone))}
	for _, entry := range undone {
		path, _ := entry.ToolArgs["path"].(string)
		lines = append(lines, fmt.Sprintf("  #%d %s(%s) — %d file di-revert", entry.ID, entry.ToolName, path, len(entry.FileChanges)))
	}
	lines = append(lines, fmt.Sprintf("\nCheckpoint tersisa: %d.", r.Manager.Count()))
	return Result{OK: true, Output: strings.Join(lines, "\n")}
}

func numericArg(value any, fallback int) int {
	switch number := value.(type) {
	case int:
		if number > 0 {
			return number
		}
	case int64:
		if number > 0 {
			return int(number)
		}
	case float64:
		if number > 0 {
			return int(number)
		}
	}
	return fallback
}
